### ISSUE TITLE
Issue with first record in ExtStore being evicted.

### ISSUE THREAD

**Issue by zipelkopf** [memcached/memcached#881]
Hi,

I'm experiencing an issue (little but annoing) when using ExtStore, that I never noticed before, problably because easy to be missed.... and I'm looking for some help please.

FYI I'm on Memcached 1.6.14 on CentOS 7 (updated to latest), dealing with records of about 1MB each, and ExtStore set in a SSD folder.

Here it is:

1. Memcached is freshly started and the ExtStore is empty.
2. Memcached starts receiving items in RAM as usual.
3.  When the RAM allocated to Memcached is about to get filled, then the ExtStore "kicks in" and starts receiving data.
4. The problem happens exactly with the record that should be placed in the ExtStore for the very first time: it gets evicted instead.
5. After that, the ExtStore works normally. 

This behavior is evident reading the "stats", showing 1 eviction.
Or running "memcache-top" (refreshing every second) and watching 1 eviction shoing up on screen in red.
 
To put it simple, the fist item ever that should be using the ExtStore, gets evicted... .after that it works fine. I wonder... is it because the ExtStore requires time to be initialised so the very first time,gets somehow lost?

I tried tuning both the Memcached server and client configuration, but without results.

Thanks a mill in advance!

**Comment by dormando** [comment]
The storage flush thread can end up sleeping for up to a second if it doesn't have to flush anything for a long time... I think the latest version has the `ext_max_sleep` as a parameter if you want to try lowering the sleep max to 100ms and see if it goes away.

If it doesn't I'll need more data as the issue template asked :) stats output and such.

**Comment by zipelkopf** [comment]
Thanks, I'm going to give it a try tomorrow and keep you posted!

P.S. I didn't know about the `ext_max_sleep`, here is the documentation I get when running with --help

Cheers!

```
[root@24980 extstore]# memcached --help
memcached 1.6.14
-p, --port=<num>          TCP port to listen on (default: 11211)
-U, --udp-port=<num>      UDP port to listen on (default: 0, off)
-s, --unix-socket=<file>  UNIX socket to listen on (disables network support)
-a, --unix-mask=<mask>    access mask for UNIX socket, in octal (default: 700)
-A, --enable-shutdown     enable ascii "shutdown" command
-l, --listen=<addr>       interface to listen on (default: INADDR_ANY)
-d, --daemon              run as a daemon
-r, --enable-coredumps    maximize core file limit
-u, --user=<user>         assume identity of <username> (only when run as root)
-m, --memory-limit=<num>  item memory in megabytes (default: 64)
-M, --disable-evictions   return error on memory exhausted instead of evicting
-c, --conn-limit=<num>    max simultaneous connections (default: 1024)
-k, --lock-memory         lock down all paged memory
-v, --verbose             verbose (print errors/warnings while in event loop)
-vv                       very verbose (also print client commands/responses)
-vvv                      extremely verbose (internal state transitions)
-h, --help                print this help and exit
-i, --license             print memcached and libevent license
-V, --version             print version and exit
-P, --pidfile=<file>      save PID in <file>, only used with -d option
-f, --slab-growth-factor=<num> chunk size growth factor (default: 1.25)
-n, --slab-min-size=<bytes> min space used for key+value+flags (default: 48)
-L, --enable-largepages  try to use large memory pages (if available)
-D <char>     Use <char> as the delimiter between key prefixes and IDs.
              This is used for per-prefix stats reporting. The default is
              ":" (colon). If this option is specified, stats collection
              is turned on automatically; if not, then it may be turned on
              by sending the "stats detail on" command to the server.
-t, --threads=<num>       number of threads to use (default: 4)
-R, --max-reqs-per-event  maximum number of requests per event, limits the
                          requests processed per connection to prevent 
                          starvation (default: 20)
-C, --disable-cas         disable use of CAS
-b, --listen-backlog=<num> set the backlog queue limit (default: 1024)
-B, --protocol=<name>     protocol - one of ascii, binary, or auto (default: auto-negotiate)
-I, --max-item-size=<num> adjusts max item size
                          (default: 1m, min: 1k, max: 1024m)
-S, --enable-sasl         turn on Sasl authentication
-F, --disable-flush-all   disable flush_all command
-X, --disable-dumping     disable stats cachedump and lru_crawler metadump
-W  --disable-watch       disable watch commands (live logging)
-Y, --auth-file=<file>    (EXPERIMENTAL) enable ASCII protocol authentication. format:
                          user:pass\nuser2:pass2\n
-e, --memory-file=<file>  (EXPERIMENTAL) mmap a file for item memory.
                          use only in ram disks or persistent memory mounts!
                          enables restartable cache (stop with SIGUSR1)
-o, --extended            comma separated list of extended options
                          most options have a 'no_' prefix to disable
   - maxconns_fast:       immediately close new connections after limit (default: enabled)
   - hashpower:           an integer multiplier for how large the hash
                          table should be. normally grows at runtime. (default starts at: 0)
                          set based on "STAT hash_power_level"
   - tail_repair_time:    time in seconds for how long to wait before
                          forcefully killing LRU tail item.
                          disabled by default; very dangerous option.
   - hash_algorithm:      the hash table algorithm
                          default is murmur3 hash. options: jenkins, murmur3, xxh3
   - no_lru_crawler:      disable LRU Crawler background thread.
   - lru_crawler_sleep:   microseconds to sleep between items
                          default is 100.
   - lru_crawler_tocrawl: max items to crawl per slab per run
                          default is 0 (unlimited)
   - read_buf_mem_limit:  limit in megabytes for connection read/response buffers.
                          do not adjust unless you have high (20k+) conn. limits.
                          0 means unlimited (default: 0)
   - no_lru_maintainer:   disable new LRU system + background thread.
   - hot_lru_pct:         pct of slab memory to reserve for hot lru.
                          (requires lru_maintainer, default pct: 20)
   - warm_lru_pct:        pct of slab memory to reserve for warm lru.
                          (requires lru_maintainer, default pct: 40)
   - hot_max_factor:      items idle > cold lru age * drop from hot lru. (default: 0.20)
   - warm_max_factor:     items idle > cold lru age * this drop from warm. (default: 2.00)
   - temporary_ttl:       TTL's below get separate LRU, can't be evicted.
                          (requires lru_maintainer, default: 61)
   - idle_timeout:        timeout for idle connections. (default: 0, no timeout)
   - slab_chunk_max:      (EXPERIMENTAL) maximum slab size in kilobytes. use extreme care. (default: 512)
   - watcher_logbuf_size: size in kilobytes of per-watcher write buffer. (default: 256)
   - worker_logbuf_size:  size in kilobytes of per-worker-thread buffer
                          read by background thread, then written to watchers. (default: 64)
   - track_sizes:         enable dynamic reports for 'stats sizes' command.
   - no_hashexpand:       disables hash table expansion (dangerous)
   - modern:              enables options which will be default in future.
                          currently: nothing
   - no_modern:           uses defaults of previous major version (1.4.x)

   - External storage (ext_*) related options (see: https://memcached.org/extstore)
   - ext_path:            file to write to for external storage.
                          ie: ext_path=/mnt/d1/extstore:1G
   - ext_page_size:       size in megabytes of storage pages. (default: 64)
   - ext_wbuf_size:       size in megabytes of page write buffers. (default: 4)
   - ext_threads:         number of IO threads to run. (default: 1)
   - ext_item_size:       store items larger than this (bytes, default 512)
   - ext_item_age:        store items idle at least this long (seconds, default: no age limit)
   - ext_low_ttl:         consider TTLs lower than this specially (default: 0)
   - ext_drop_unread:     don't re-write unread values during compaction (default: disabled)
   - ext_recache_rate:    recache an item every N accesses (default: 2000)
   - ext_compact_under:   compact when fewer than this many free pages
                          (default: 1/4th of the assigned storage)
   - ext_drop_under:      drop COLD items when fewer than this many free pages
                          (default: 1/4th of the assigned storage)
   - ext_max_frag:        max page fragmentation to tolerate (default: 0.80)
   - slab_automove_freeratio: ratio of memory to hold free as buffer.
                          (see doc/storage.txt for more info, default: 0.010)
-N, --napi_ids            number of napi ids. see doc/napi_ids.txt for more details

```

**Comment by zipelkopf** [comment]
I couldn't resist and I give it ia try now ... unfortunatelly it doesn't like the "ext_max_sleep".
I tried with both 100 and 100ms setting... just in case.

I will post more details as per template tomorrow.

Thanks again!

```
Mar 10 23:07:44 24980.18391 systemd[1]: Started Memcached.
Mar 10 23:07:44 24980.18391 memcached[19520]: Illegal suboption "ext_max_sleep=100"
Mar 10 23:07:44 24980.18391 systemd[1]: memcached.service: main process exited, code=exited, status=1/FAILURE
Mar 10 23:07:44 24980.18391 systemd[1]: Unit memcached.service entered failed state.
Mar 10 23:07:44 24980.18391 systemd[1]: memcached.service failed.
```

**Comment by dormando** [comment]
Ah sorry, that's still sitting in the `next` tree. If you want to try building from github's `next` branch it should be in there. Compiling just requires an `./autogen.sh` prestep and libtools/automake/autoconf installed (see BUILD/README).

The next release should be out soon anyway.

**Comment by zipelkopf** [comment]
I found a reference to the ext_max_sleep in the conversation #432 ...but I'm not a "guru" here so I feel more confident in waiting for your tag, but I will put together anyway a report in case it might help you.

**Comment by dormando** [comment]
You don't need to do anything special other than check out the `next` branch. The change is already in there. Anyway if you can't, probably next week it'll be tagged.

**Comment by zipelkopf** [comment]
@dormando I built the `next` release and it was easy as you said indeed, thanks!

**To Reproduce**

Unfortunatelly the problem persists, even with the 'ext_max_sleep' reduced to 100... and I also found another issue (please don't hate me!!!) ...unless it is by design?: if I `restart` memcached , then the extstore never receives anything (the flash marks 0 'zero' size on the filesystem) and items that are supposed to go to the extstore are always evicted. 

To make it working, I must first `stop` memcached, then manually delete the flash file even if the filesize is 0 'zero' (i.e. rm /vstore/extstore/flash), then `start` memcached, then it works... but still the first item that is supposed to go to the extstore gets evicted.

Please find below the details by the template.
_Just let me know if you need more, happy to help!_

**System Information**
 - OS/Distro: CentOS 7
 - Version of OS/distro: 7.9.2009
 - Version of memcached: "next" branch downloaded today
 - Hardware detail: 2 CPU (Intel), 3 GB, SSD 

**Memcached configuration**
Please note I use this configuration just for testing, so to quicly add only a few large items (of few MB each) to saturate the cachesize in RAM and attempt using the ExtStore.
FYI, I push items to memcached manually 1 by 1 during testing, only after the previous has completed, so to have a visual feedback of when the eviction happened.

> PORT="11211"
> USER="memcached"
> MAXCONN="1024"
> CACHESIZE="40"
> OPTIONS="-l 127.0.0.1 -U 0 -I 20m -o ext_page_size=128,ext_path=/vstore/extstore/flash:1G,ext_wbuf_size=32,ext_threads=2,ext_item_size=512,ext_max_sleep=100"

### stats

> STAT pid 6393
> STAT uptime 199
> STAT time 1647035973
> STAT version UNKNOWN
> STAT libevent 2.0.21-stable
> STAT pointer_size 64
> STAT rusage_user 57.128398
> STAT rusage_system 53.201620
> STAT max_connections 1024
> STAT curr_connections 1
> STAT total_connections 176
> STAT rejected_connections 0
> STAT connection_structures 3
> STAT response_obj_oom 0
> STAT response_obj_count 1
> STAT response_obj_bytes 65536
> STAT read_buf_count 8
> STAT read_buf_bytes 131072
> STAT read_buf_bytes_free 49152
> STAT read_buf_oom 0
> STAT reserved_fds 20
> STAT cmd_get 8
> STAT cmd_set 5
> STAT cmd_flush 0
> STAT cmd_touch 0
> STAT cmd_meta 0
> STAT get_hits 0
> STAT get_misses 8
> STAT get_expired 0
> STAT get_flushed 0
> STAT get_extstore 0
> STAT get_aborted_extstore 0
> STAT get_oom_extstore 0
> STAT recache_from_extstore 0
> STAT miss_from_extstore 0
> STAT badcrc_from_extstore 0
> STAT delete_misses 0
> STAT delete_hits 0
> STAT incr_misses 0
> STAT incr_hits 0
> STAT decr_misses 0
> STAT decr_hits 0
> STAT cas_misses 0
> STAT cas_hits 0
> STAT cas_badval 0
> STAT touch_hits 0
> STAT touch_misses 0
> STAT store_too_large 0
> STAT store_no_memory 0
> STAT auth_cmds 0
> STAT auth_errors 0
> STAT bytes_read 49098701
> STAT bytes_written 487877
> STAT limit_maxbytes 41943040
> STAT accepting_conns 1
> STAT listen_disabled_num 0
> STAT time_in_listen_disabled_us 0
> STAT threads 4
> STAT conn_yields 0
> STAT hash_power_level 16
> STAT hash_bytes 524288
> STAT hash_is_expanding 0
> STAT slab_reassign_rescues 0
> STAT slab_reassign_chunk_rescues 18
> STAT slab_reassign_evictions_nomem 0
> STAT slab_reassign_inline_reclaim 0
> STAT slab_reassign_busy_items 0
> STAT slab_reassign_busy_deletes 0
> STAT slab_reassign_running 0
> STAT slabs_moved 16
> STAT lru_crawler_running 0
> STAT lru_crawler_starts 4
> STAT lru_maintainer_juggles 591
> STAT malloc_fails 0
> STAT log_worker_dropped 0
> STAT log_worker_written 0
> STAT log_watcher_skipped 0
> STAT log_watcher_sent 0
> STAT log_watchers 0
> STAT extstore_compact_lost 0
> STAT extstore_compact_rescues 0
> STAT extstore_compact_skipped 0
> STAT extstore_page_allocs 0
> STAT extstore_page_evictions 0
> STAT extstore_page_reclaims 0
> STAT extstore_pages_free 7
> STAT extstore_pages_used 1
> STAT extstore_objects_evicted 0
> STAT extstore_objects_read 0
> STAT extstore_objects_written 0
> STAT extstore_objects_used 0
> STAT extstore_bytes_evicted 0
> STAT extstore_bytes_written 0
> STAT extstore_bytes_read 0
> STAT extstore_bytes_used 0
> STAT extstore_bytes_fragmented 134217728
> STAT extstore_limit_maxbytes 1073741824
> STAT extstore_io_queue 0
> STAT unexpected_napi_ids 0
> STAT round_robin_fallback 0
> STAT bytes 24548992
> STAT curr_items 3
> STAT total_items 5
> STAT slab_global_page_pool 9
> STAT expired_unfetched 0
> STAT evicted_unfetched 2
> STAT evicted_active 0
> STAT evictions 2
> STAT reclaimed 0
> STAT crawler_reclaimed 0
> STAT crawler_items_checked 6
> STAT lrutail_reflocked 0
> STAT moves_to_cold 5
> STAT moves_to_warm 0
> STAT moves_within_lru 0
> STAT direct_reclaims 2
> STAT lru_bumps_dropped 0
> END

### stats settings

> STAT maxbytes 41943040
> STAT maxconns 1024
> STAT tcpport 11211
> STAT udpport 0
> STAT inter 127.0.0.1
> STAT verbosity 0
> STAT oldest 0
> STAT evictions on
> STAT domain_socket NULL
> STAT umask 700
> STAT shutdown_command no
> STAT growth_factor 1.25
> STAT chunk_size 48
> STAT num_threads 4
> STAT num_threads_per_udp 4
> STAT stat_key_prefix :
> STAT detail_enabled no
> STAT reqs_per_event 20
> STAT cas_enabled yes
> STAT tcp_backlog 1024
> STAT binding_protocol auto-negotiate
> STAT auth_enabled_sasl no
> STAT auth_enabled_ascii no
> STAT item_size_max 20971520
> STAT maxconns_fast yes
> STAT hashpower_init 0
> STAT slab_reassign yes
> STAT slab_automove 1
> STAT slab_automove_ratio 0.80
> STAT slab_automove_window 30
> STAT slab_chunk_max 524288
> STAT lru_crawler yes
> STAT lru_crawler_sleep 100
> STAT lru_crawler_tocrawl 0
> STAT tail_repair_time 0
> STAT flush_enabled yes
> STAT dump_enabled yes
> STAT hash_algorithm murmur3
> STAT lru_maintainer_thread yes
> STAT lru_segmented yes
> STAT hot_lru_pct 20
> STAT warm_lru_pct 40
> STAT hot_max_factor 0.20
> STAT warm_max_factor 2.00
> STAT temp_lru no
> STAT temporary_ttl 61
> STAT idle_timeout 0
> STAT watcher_logbuf_size 262144
> STAT worker_logbuf_size 65536
> STAT read_buf_mem_limit 0
> STAT track_sizes no
> STAT inline_ascii_response no
> STAT ext_item_size 512
> STAT ext_item_age 4294967295
> STAT ext_low_ttl 0
> STAT ext_recache_rate 2000
> STAT ext_wbuf_size 33554432
> STAT ext_compact_under 2
> STAT ext_drop_under 2
> STAT ext_max_sleep 100
> STAT ext_max_frag 0.80
> STAT slab_automove_freeratio 0.010
> STAT ext_drop_unread no
> STAT num_napi_ids (null)
> STAT memory_file (null)
> END

### stats items

> STAT items:13:number 1
> STAT items:13:number_hot 0
> STAT items:13:number_warm 0
> STAT items:13:number_cold 1
> STAT items:13:age_hot 0
> STAT items:13:age_warm 0
> STAT items:13:age 342
> STAT items:13:mem_requested 1264
> STAT items:13:evicted 0
> STAT items:13:evicted_nonzero 0
> STAT items:13:evicted_time 0
> STAT items:13:outofmemory 0
> STAT items:13:tailrepairs 0
> STAT items:13:reclaimed 0
> STAT items:13:expired_unfetched 0
> STAT items:13:evicted_unfetched 0
> STAT items:13:evicted_active 0
> STAT items:13:crawler_reclaimed 0
> STAT items:13:crawler_items_checked 5
> STAT items:13:lrutail_reflocked 0
> STAT items:13:moves_to_cold 1
> STAT items:13:moves_to_warm 0
> STAT items:13:moves_within_lru 0
> STAT items:13:direct_reclaims 0
> STAT items:13:hits_to_hot 0
> STAT items:13:hits_to_warm 0
> STAT items:13:hits_to_cold 0
> STAT items:13:hits_to_temp 0
> STAT items:39:number 2
> STAT items:39:number_hot 0
> STAT items:39:number_warm 0
> STAT items:39:number_cold 2
> STAT items:39:age_hot 0
> STAT items:39:age_warm 0
> STAT items:39:age 291
> STAT items:39:mem_requested 24547728
> STAT items:39:evicted 2
> STAT items:39:evicted_nonzero 2
> STAT items:39:evicted_time 35
> STAT items:39:outofmemory 0
> STAT items:39:tailrepairs 0
> STAT items:39:reclaimed 0
> STAT items:39:expired_unfetched 0
> STAT items:39:evicted_unfetched 2
> STAT items:39:evicted_active 0
> STAT items:39:crawler_reclaimed 0
> STAT items:39:crawler_items_checked 4
> STAT items:39:lrutail_reflocked 0
> STAT items:39:moves_to_cold 4
> STAT items:39:moves_to_warm 0
> STAT items:39:moves_within_lru 0
> STAT items:39:direct_reclaims 2
> STAT items:39:hits_to_hot 0
> STAT items:39:hits_to_warm 0
> STAT items:39:hits_to_cold 0
> STAT items:39:hits_to_temp 0
> END

### stats slabs

> STAT 5:chunk_size 240
> STAT 5:chunks_per_page 4369
> STAT 5:total_pages 1
> STAT 5:total_chunks 4369
> STAT 5:used_chunks 2
> STAT 5:free_chunks 4367
> STAT 5:free_chunks_end 0
> STAT 5:get_hits 0
> STAT 5:cmd_set 0
> STAT 5:delete_hits 0
> STAT 5:incr_hits 0
> STAT 5:decr_hits 0
> STAT 5:cas_hits 0
> STAT 5:cas_badval 0
> STAT 5:touch_hits 0
> STAT 13:chunk_size 1480
> STAT 13:chunks_per_page 708
> STAT 13:total_pages 2
> STAT 13:total_chunks 1416
> STAT 13:used_chunks 1
> STAT 13:free_chunks 1415
> STAT 13:free_chunks_end 0
> STAT 13:get_hits 0
> STAT 13:cmd_set 1
> STAT 13:delete_hits 0
> STAT 13:incr_hits 0
> STAT 13:decr_hits 0
> STAT 13:cas_hits 0
> STAT 13:cas_badval 0
> STAT 13:touch_hits 0
> STAT 36:chunk_size 252696
> STAT 36:chunks_per_page 4
> STAT 36:total_pages 3
> STAT 36:total_chunks 12
> STAT 36:used_chunks 2
> STAT 36:free_chunks 10
> STAT 36:free_chunks_end 0
> STAT 36:get_hits 0
> STAT 36:cmd_set 0
> STAT 36:delete_hits 0
> STAT 36:incr_hits 0
> STAT 36:decr_hits 0
> STAT 36:cas_hits 0
> STAT 36:cas_badval 0
> STAT 36:touch_hits 0
> STAT 39:chunk_size 524288
> STAT 39:chunks_per_page 2
> STAT 39:total_pages 25
> STAT 39:total_chunks 50
> STAT 39:used_chunks 46
> STAT 39:free_chunks 4
> STAT 39:free_chunks_end 0
> STAT 39:get_hits 0
> STAT 39:cmd_set 4
> STAT 39:delete_hits 0
> STAT 39:incr_hits 0
> STAT 39:decr_hits 0
> STAT 39:cas_hits 0
> STAT 39:cas_badval 0
> STAT 39:touch_hits 0
> STAT active_slabs 4
> STAT total_malloced 41943040
> END

**Comment by zipelkopf** [comment]
P.S.
I really hope this can help, but if I add `ext_item_age=0` to the configuration, everything work like a charm!

The --help says: "_...store items idle at least this long (seconds, default: no age limit)"

Can you please explain a bit better and why it works when setting to '0' zero?
When and how should I used? Any advice please?

THANKS!

**Comment by dormando** [comment]
in normal mode extstore flushes items once memory is full. with `item_age` set it'll flush items once they've been "idle" at least that long, with 0 that means immediate.

Your memory limit is 41 megabytes? that's pretty small. do you still have the issue with something a bit bigger?

**Comment by zipelkopf** [comment]
Thanks for getting back to me quickly.

So , if I understand well, an item larger than `ext_item_size = 512` will never sit in RAM but always in the Flash drive when `ext_item_age = 0` ...so I could use the RAM only for items smaller than 'ext_item_age = 512' (or whatever value I set).
If that's the case, then is great! :-)

> Your memory limit is 41 megabytes? that's pretty small. do you still have the issue with something a bit bigger?

If I do not set `ext_item_age = 0`, the issue happens also with larger memory (i.e. 128 MB), I simply set a lower value of 40 MB to test this behavior quickly.

Just drop me a message if you spot anything to fix and I will be happy to run some additional tests for you, now that I know how to build memcached from source 👍 
 
P.S.
Using the setting `ext_item_age = 0` in the Memcached 1.6.14 works perfectly as well, with no evicted items, I have just tried it.

Enjoy the weekend!

**Comment by dormando** [comment]
re: first part, yes that's how it works.

I think 128M is still pretty low as far as memcached flash ecosystem goes. I'm still going to leave this issue open until I get some time to poke at it.

**Comment by dormando** [comment]
https://github.com/memcached/memcached/pull/923 - I've finally made an adjustment to the flushing algorithm. This should be a halfway point between the larger rewrite I would like to do, and something that just has workable defaults.

The original automove algorithm probably only works well with huge amounts of memory and a relatively slow fill rate. While 128mb is probably still on the too-low end, it might actually work now at least. 512mb+ are probably safe.

**Comment by dormando** [comment]
fix released in 1.6.17

**Comment by zipelkopf** [comment]
@dormando thanks a mill!
I will be doing an upgrade soon 👍

**Comment by dormando** [comment]
@zipelkopf thanks! Please open a new issue if you're still seeing a problem
